src/query/service/src/interpreters/interpreter_copy.rs

```diff
@@ -454,8 +407,56 @@ impl Interpreter for CopyInterpreter {
         }
 
         match &self.plan {
-            CopyPlan::IntoTable(plan) => self.build_copy_into_table_pipeline(plan).await,
-
+            CopyPlan::IntoTable(plan) => {
+                if plan.enable_distributed {
+                    let (distributed_plan_op, empty_files) =
+                        self.transform_copy_plan_distributed(plan).await?;
+                    if empty_files {
+                        return self.build_copy_into_table_pipeline(plan).await;
+                    }
+                    let distributed_plan = distributed_plan_op.unwrap();
+
+                    // add exchange plan node to enable distributed
+                    // TODO(leiysky): we reuse the id of exchange here,
+                    // which is not correct. We should generate a new id for insert.
+                    let exchange_plan = PhysicalPlan::Exchange(Exchange {
+                        plan_id: 0,
+                        input: Box::new(distributed_plan.clone()),
+                        kind: FragmentKind::Merge,
+                        keys: Vec::new(),
+                    });
+                    let mut build_res =
+                        build_distributed_pipeline(&self.ctx, &exchange_plan, false).await?;
+                    match distributed_plan {
+                        PhysicalPlan::DistributedCopyIntoTable(distributed_plan) => {
+                            let catalog = self.ctx.get_catalog(&distributed_plan.catalog_name)?;
+                            let to_table =
+                                catalog.get_table_by_info(&distributed_plan.table_info)?;
+                            let copied_files = CopyInterpreter::upsert_copied_files_request(
+                                self.ctx.clone(),
+                                to_table.clone(),
+                                distributed_plan.stage_table_info.stage_info.clone(),
+                                distributed_plan.files.clone(),
+                                distributed_plan.force,
+                            )?;
+                            let mut overwrite_ = false;
+                            if let CopyIntoTableMode::Insert { overwrite } = plan.write_mode {
+                                overwrite_ = overwrite;
+                            }
+                            to_table.commit_insertion(
+                                self.ctx.clone(),
+                                &mut build_res.main_pipeline,
+                                copied_files,
+                                overwrite_,
+                            )?
+                        }
+                        _ => unreachable!(),
```

Comment:
can we avoid this `unreachable` match branch, by letting `transform_copy_plan_distributed` not return the enum but the struct that embedded in the enum variant?